File: GameEngine_IOS/GameEngine/Shaders/Terrains/TerrainShaderManager.m
#import "TerrainShaderManager.h"

@implementation TerrainShaderManager
{
@private
    
    // Uniform index.
    enum
    {
        TRANSFORMATION_MATRIX,
        VIEW_MATRIX,
        PROJECTION_MATRIX,
        
        LIGHT_POSITION,
        LIGHT_COLOR,
        SHINE_DAMPER,
        REFLECTIVITY,
        SKY_COLOR,
        BACKGROUND_TEXTURE,
        MUD_TEXTURE,
        GRASS_TEXTURE,
        PATH_TEXTURE,
        WEIGHT_TEXTURE,
        NUM_UNIFORMS
    };
    GLint uniforms[NUM_UNIFORMS];
    
}

/**
 * Initializor of the game shader where the vertex and fragment shader of
 * the game engine are loaded
 *
 */
- (id)init {
    self = [super init: @"terrain_vertex_shader" : @"terrain_fragment_shader"];
    return self;
}

/**
 * Called to bind the attributes to the program shader
 */
- (void) bindAttributes {
    [super bindAttribute : LOCATION_ATTR_ID: "position"];
    [super bindAttribute : TEX_COORDINATE_ATTR_ID: "textureCoords"];
    [super bindAttribute : NORMAL_VECTOR_ATTR_ID: "normal"];
}

/**
 * Called to ensure that all the shader managers get their uniform locations
 */
- (void) getAllUniformLocations {
    @autoreleasepool {
        NSDictionary * uniformsDic = [NSDictionary dictionaryWithObjectsAndKeys:
                                      @"projectionMatrix", [NSNumber numberWithInt:PROJECTION_MATRIX],
                                      @"viewMatrix", [NSNumber numberWithInt: VIEW_MATRIX],
                                      @"transformationMatrix", [NSNumber numberWithInt: TRANSFORMATION_MATRIX],
                                      @"lightPosition", [NSNumber numberWithInt:LIGHT_POSITION],
                                      @"lightColor", [NSNumber numberWithInt:LIGHT_COLOR],
                                      @"shineDamper", [NSNumber numberWithInt:SHINE_DAMPER],
                                      @"reflectivity", [NSNumber numberWithInt:REFLECTIVITY],
                                      @"skyColor", [NSNumber numberWithInt:SKY_COLOR],
                                      @"backgroundTexture", [NSNumber numberWithInt: BACKGROUND_TEXTURE],
                                      @"mudTexture",[NSNumber numberWithInt: MUD_TEXTURE],
                                      @"grassTexture",[NSNumber numberWithInt: GRASS_TEXTURE],
                                      @"pathTexture",[NSNumber numberWithInt: PATH_TEXTURE],
                                      @"weightMapTexture",[NSNumber numberWithInt: WEIGHT_TEXTURE],
                                      nil];
        for(id key in uniformsDic) {
            int iKey = [key intValue];
            const char *shaderVariable = [(uniformsDic[key]) UTF8String];
            uniforms[iKey] = [super getUniformLocation: shaderVariable];
            if(uniforms[iKey] < 0) {
                NSLog(@"Problems getting %@'s location in the shader", uniformsDic[key]);
            }
        }
    }
}

- (void) connectTextureUnits {
    [super loadInt: uniforms[BACKGROUND_TEXTURE] : TEXTURE_UNIT0];
    [super loadInt: uniforms[MUD_TEXTURE] : TEXTURE_UNIT1];
    [super loadInt: uniforms[GRASS_TEXTURE] : TEXTURE_UNIT2];
    [super loadInt: uniforms[PATH_TEXTURE] : TEXTURE_UNIT3];
    [super loadInt: uniforms[WEIGHT_TEXTURE] : TEXTURE_UNIT4];
    
}

/**
 * Load the projection matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadProjectionMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[PROJECTION_MATRIX]: matrix];
}

/**
 * Load the view matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadViewMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[VIEW_MATRIX]: matrix];
}


/**
 * Load the transformation matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadTransformationMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[TRANSFORMATION_MATRIX]: matrix];
}

/**
 * Put passes the information of the light to the
 * Shader program
 *
 * @param light the light to load in the shader program
 */
- (void) loadLight : (Light*) light {
    [super loadVector: uniforms[LIGHT_POSITION] : [light position]];
    [super loadVector: uniforms[LIGHT_COLOR] : [light color]];
}

/**
 * Load the values of the specular light in the fragment shader
 *
 * @param damper		The damper of the specular lighting
 * @param reflectivity	The reflectivity of the material
 */
- (void) loadShineVariables : (float) damper : (float) reflectivity {
    [super loadFloat : uniforms[SHINE_DAMPER] : damper];
    [super loadFloat : uniforms[REFLECTIVITY] : reflectivity];
}


/**
 * Load the color of the sky in order to simulate fog
 *
 * @param skyColor
 * 			Color of the sky
 */
- (void)  loadSkyColor : (Vector3f*) skyColor {
    [super loadVector: uniforms[SKY_COLOR] : skyColor];
}

@end

File: GameEngine_IOS/GameEngine/Shaders/Terrains/TerrainShaderManager.swift
import Foundation

/**
 * Manager of the shader files that are going to be load to render the 3D
 * Terrain
 */
public class TerrainShaderManager : ShaderManager {
    
    /**
     * Initializor of the game shader where the vertex and fragment shader of
     * the game engine are loaded
     *
     */
    public init() {
        super.init(vertexFile: "terrain_vertex_shader" , fragmentFile: "terrain_fragment_shader", numberOfUniforms: TTerrainUniform.numOfTerrainLocations.rawValue)
    }
    
    
    /**
     * Called to bind the attributes to the program shader
     */
    override internal func getAttributes() -> Dictionary<Int, String>! {
        let attributesDic : Dictionary<Int, String> = [
            TTerrainAttribute.position.rawValue : "\(TTerrainAttribute.position)",
            TTerrainAttribute.textureCoords.rawValue : "\(TTerrainAttribute.textureCoords)",
            TTerrainAttribute.normal.rawValue : "\(TTerrainAttribute.normal)"
        ];
        return attributesDic;
    }
    
    /**
     * Called to provide the uniform locations that are going to get bound to the shader
     */
    override internal func getUniformLocations() -> Dictionary<NSInteger, String>! {
        let uniformsDic : Dictionary<NSInteger, String> = [
            TTerrainUniform.projectionMatrix.rawValue : "\(TTerrainUniform.projectionMatrix)",
            TTerrainUniform.viewMatrix.rawValue : "\(TTerrainUniform.viewMatrix)",
            TTerrainUniform.transformationMatrix.rawValue : "\(TTerrainUniform.transformationMatrix)",
            TTerrainUniform.lightPosition.rawValue : "\(TTerrainUniform.lightPosition)",
            TTerrainUniform.lightColor.rawValue : "\(TTerrainUniform.lightColor)",
            TTerrainUniform.shineDamper.rawValue : "\(TTerrainUniform.shineDamper)",
            TTerrainUniform.reflectivity.rawValue : "\(TTerrainUniform.reflectivity)",
            TTerrainUniform.skyColor.rawValue : "\(TTerrainUniform.skyColor)",
            TTerrainUniform.backgroundTexture.rawValue : "\(TTerrainUniform.backgroundTexture)",
            TTerrainUniform.mudTexture.rawValue : "\(TTerrainUniform.mudTexture)",
            TTerrainUniform.grassTexture.rawValue : "\(TTerrainUniform.grassTexture)",
            TTerrainUniform.pathTexture.rawValue : "\(TTerrainUniform.pathTexture)",
            TTerrainUniform.weightMapTexture.rawValue : "\(TTerrainUniform.weightMapTexture)"
        ]
        
        return uniformsDic;
    }
    
    /**
     * Associate the shader variables with textures that were defined in the
     * bind of textures
     */
    public func connectTextureUnits() {
        super.loadInt(uniforms[TTerrainUniform.backgroundTexture.rawValue], value: TTerrainTexture.TEXTURE_UNIT0.rawValue);
        super.loadInt(uniforms[TTerrainUniform.mudTexture.rawValue], value: TTerrainTexture.TEXTURE_UNIT1.rawValue);
        super.loadInt(uniforms[TTerrainUniform.grassTexture.rawValue], value: TTerrainTexture.TEXTURE_UNIT2.rawValue);
        super.loadInt(uniforms[TTerrainUniform.pathTexture.rawValue], value: TTerrainTexture.TEXTURE_UNIT3.rawValue);
        super.loadInt(uniforms[TTerrainUniform.weightMapTexture.rawValue], value: TTerrainTexture.TEXTURE_UNIT4.rawValue);
    }
    
    /**
     * Load the projection matrix
     *
     * @param matrix
     *            the matrix to be loaded
     */
    public func loadProjectionMatrix (matrix : GLTransformation) {
        super.loadMatrix(uniforms[TTerrainUniform.projectionMatrix.rawValue], matrix: matrix);
    }
    
    
    /**
     * Load the view matrix
     *
     * @param matrix
     *            the matrix to be loaded
     */
    public func loadViewMatrix (matrix : GLTransformation) {
        super.loadMatrix(uniforms[TTerrainUniform.viewMatrix.rawValue], matrix: matrix);
    }
    
    /**
     * Load the transformation matrix
     *
     * @param matrix
     *            the matrix to be loaded
     */
    public func loadTransformationMatrix(matrix : GLTransformation) {
        super.loadMatrix(uniforms[TTerrainUniform.transformationMatrix.rawValue], matrix: matrix);
    }
    
    /**
     * Put passes the information of the light to the
     * Shader program
     *
     * @param light the light to load in the shader program
     */
    public func loadLight(light : Light) {
        super.loadVector(uniforms[TTerrainUniform.lightPosition.rawValue], vector: light.position);
        super.loadColorRGB(uniforms[TTerrainUniform.lightColor.rawValue], color: light.color);
    }
    
    /**
     * Load the values of the specular light in the fragment shader
     *
     * @param damper		The damper of the specular lighting
     * @param reflectivity	The reflectivity of the material
     */
    public func loadShineVariables(damper : Float, reflectivity : Float) {
        super.loadFloat(uniforms[TTerrainUniform.shineDamper.rawValue], value: damper);
        super.loadFloat(uniforms[TTerrainUniform.reflectivity.rawValue], value : reflectivity);
    }
    
    
    /**
     * Load the color of the sky in order to simulate fog
     *
     * @param skyColor
     * 			Color of the sky
     */
    public func  loadSkyColor(skyColor : ColorRGBA) {
        super.loadColorRGBA(uniforms[TTerrainUniform.skyColor.rawValue], color: skyColor);
    }
}
